client/src/hooks/queryParam/index.ts

```diff
@@ -0,0 +1,90 @@
+import { useRouter } from 'next/router';
+import { useCallback, useState, useTransition } from 'react';
+import { useDebounce, useEffectOnceWhen } from 'rooks';
+
+interface UseQueryParamOptions<T, F = T> {
+  /**
+   * Format the param before it's JSON.stringified
+   * Useful to store less information in the query params
+   * and avoid clutter
+   */
+  formatParam?: (value: T) => F;
+}
+
+const serialize = <T>(value: T): string => {
+  if (value === undefined) return undefined;
+
+  return encodeURIComponent(JSON.stringify(value));
+};
+
+const parse = <T>(value: string): T => {
+  if (value === undefined) return undefined;
+
+  return JSON.parse(decodeURIComponent(value));
+};
+
+const useQueryParam = <T, F = T>(
+  name: string,
+  defaultValue?: T | (() => T),
+  { formatParam }: UseQueryParamOptions<T, F> = {},
+) => {
+  const { query, isReady, replace } = useRouter();
+
+  const [value, setValue] = useState(defaultValue);
+
+  const queryValue = query[name] as string;
+
+  const [isDoneSettingFromQuery, setIsDoneSettingFromQuery] = useState(false);
+  useEffectOnceWhen(() => {
+    const newValue = parse<T>(query[name] as string);
+    if (typeof defaultValue === 'object' || typeof newValue === 'object') {
+      setValue((value) => ({ ...value, ...newValue }));
+    } else {
+      setValue(newValue);
+    }
+    setIsDoneSettingFromQuery(true);
+  }, isReady);
+
+  const [isPending, startTransition] = useTransition();
+
+  const setQueryParam = useCallback(
+    (value: T) => {
+      if (!isReady) return;
+      if (!value) {
+        // delete the old value from the query params
+        if (queryValue) {
+          const { [name]: _currentValue, ...restQuery } = query;
+          replace({ query: restQuery }, undefined, { shallow: true });
+        }
+        return;
+      }
+
+      replace({
+        query: {
+          ...query,
+          [name]: serialize(formatParam ? formatParam(value) : value),
+        },
+      });
+    },
+    [formatParam, isReady, name, query, queryValue, replace],
+  );
+
+  const setDebouncedQueryParam = useDebounce(setQueryParam, 500);
```

Comment:
Could you reduce the time of the debounce to 100 ms?? Also, it's better if this time is on the top of the doc in a constant.